// jcatalog-react-reference-search/lib/components/ReferenceSearchDialogProps/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

exports.default = {

  /**
   * Title of the reference search dialog
   */
  title: _react.PropTypes.string.isRequired,

  /**
   * Array of search columns, where object is:
   *
   * @name - field name (f.e. customerId)
   * @label - title of field name (usually translation of the field name f.e. 'Customer Id')
   * @inputComponent - instance of the react component that will be render instead of default input
   * The component should support the next props:
   * *onChange - function that should be called on in input change
   * *value - input value
   * *ref - ref should be pointed to target input to provide focus
   * *id - id of the input component to provide correct label work
   *
   * The component should provide method focus that targets real input, it can look like:
   *   focus() {
   *   let element = this.refs[this.props.id];
   *   if (element) {
   *     setTimeout(() => element.focus(), 300);
   *   }
   * }
   * Check out TestReferenceSearchDialogInputComponent.react.js for example
   *
   *
   */
  searchFields: _react.PropTypes.arrayOf(_react.PropTypes.shape({
    name: _react.PropTypes.string.isRequired,
    label: _react.PropTypes.string.isRequired,
    inputComponent: _react.PropTypes.func
  })).isRequired,

  /**
   * Array of search result columns, where object is:
   *
   * @name - field name (f.e. customerId)
   * @label - title of field name (usually translation of the field name f.e. 'Customer Id')
   * @sortable - true/false
   */
  resultFields: _react.PropTypes.arrayOf(_react.PropTypes.shape({
    name: _react.PropTypes.string.isRequired,
    label: _react.PropTypes.string.isRequired,
    sortable: _react.PropTypes.bool
  })).isRequired
};
