// jcatalog-react-reference-search/lib/components/ContextInitContainer/ContextInitContainer.react.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _jcatalogI18n = require('jcatalog-i18n');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ContextInitContainer = (_temp2 = _class = function (_React$Component) {
  _inherits(ContextInitContainer, _React$Component);

  function ContextInitContainer() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ContextInitContainer);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ContextInitContainer.__proto__ || Object.getPrototypeOf(ContextInitContainer)).call.apply(_ref, [this].concat(args))), _this), _this.getChildContext = function () {
      if (!_this.context.i18n) {
        _this.context.i18n = new _jcatalogI18n.I18nManager(_this.props.locale, null, _this.props.formatPatterns);
      }

      return {
        i18n: _this.context.i18n
      };
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ContextInitContainer, [{
    key: 'render',
    value: function render() {
      return this.props.children;
    }
  }]);

  return ContextInitContainer;
}(_react2.default.Component), _class.propTypes = {
  locale: _react2.default.PropTypes.string,
  formatPatterns: _react2.default.PropTypes.object,
  children: _react2.default.PropTypes.node
}, _class.contextTypes = {
  locale: _react2.default.PropTypes.string,
  i18n: _react2.default.PropTypes.object
}, _class.childContextTypes = {
  i18n: _react2.default.PropTypes.object.isRequired
}, _class.defaultProps = {
  formatPatterns: {
    'en': {
      'datePattern': 'MM/dd/yyyy',
      'integerPattern': '#,##0',
      'numberPattern': '#,##0.00',
      'numberDecimalSeparator': '.',
      'numberGroupingSeparator': ',',
      'numberGroupingSeparatorUse': true
    },
    de: {
      datePattern: "dd.MM.yyyy",
      integerPattern: "#,##0",
      numberPattern: "#,##0.00",
      numberDecimalSeparator: ",",
      numberGroupingSeparator: ".",
      numberGroupingSeparatorUse: true
    }
  },

  locale: 'en'
}, _temp2);
exports.default = ContextInitContainer;
