// jcatalog-react-reference-search/lib/components/ClassificationInput/ClassificationInput.react.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ReferenceAutocomplete = require('../ReferenceAutocomplete');

var _ReferenceAutocomplete2 = _interopRequireDefault(_ReferenceAutocomplete);

var _ReferenceInputBaseProps = require('../ReferenceInputBaseProps');

var _ReferenceInputBaseProps2 = _interopRequireDefault(_ReferenceInputBaseProps);

var _ReactSelectSpecificProps = require('../ReactSelectSpecificProps');

var _ReactSelectSpecificProps2 = _interopRequireDefault(_ReactSelectSpecificProps);

var _ClassificationService = require('../../service/ClassificationService');

var _ClassificationService2 = _interopRequireDefault(_ClassificationService);

var _ServiceRegistryValidator = require('../ServiceRegistryValidator');

var _ServiceRegistryValidator2 = _interopRequireDefault(_ServiceRegistryValidator);

var _refreshValueDecorator = require('../../utils/refreshValueDecorator.react');

var _refreshValueDecorator2 = _interopRequireDefault(_refreshValueDecorator);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SERVICE_NAME = 'classification';

var loadClassification = function loadClassification(serviceRegistry) {
  return function (id) {
    return new _ClassificationService2.default(serviceRegistry(SERVICE_NAME).url).getClassification(id);
  };
};

var ClassificationInput = (_dec = (0, _refreshValueDecorator2.default)('classificationId', '_objectLabel', loadClassification), _dec(_class = (_temp = _class2 = function (_React$Component) {
  _inherits(ClassificationInput, _React$Component);

  function ClassificationInput() {
    _classCallCheck(this, ClassificationInput);

    return _possibleConstructorReturn(this, (ClassificationInput.__proto__ || Object.getPrototypeOf(ClassificationInput)).apply(this, arguments));
  }

  _createClass(ClassificationInput, [{
    key: 'render',
    value: function render() {
      var serviceUrl = this.props.serviceRegistry(SERVICE_NAME).url;
      var classificationService = new _ClassificationService2.default(serviceUrl);
      var autocompleteProps = {
        autocompleteAction: function autocompleteAction(query) {
          return classificationService.getClassifications().then(function (response) {
            return {
              options: response.body,
              complete: true
            };
          });
        },
        labelProperty: '_objectLabel',
        valueProperty: 'classificationId',
        reactSelectSpecificProps: this.props.reactSelectSpecificProps
      };
      return _react2.default.createElement(_ReferenceAutocomplete2.default, _extends({}, this.props, autocompleteProps));
    }
  }]);

  return ClassificationInput;
}(_react2.default.Component), _class2.propTypes = _extends({}, _ReferenceInputBaseProps2.default, {
  reactSelectSpecificProps: _react2.default.PropTypes.shape(_ReactSelectSpecificProps2.default),
  serviceRegistry: (0, _ServiceRegistryValidator2.default)(SERVICE_NAME)
}), _temp)) || _class);
exports.default = ClassificationInput;
