// jcatalog-react-reference-search/lib/components/TermReferenceSearchDialog/TermReferenceSearchDialog.SCOPE.react.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _jcatalogShowroom = require('jcatalog-showroom');

var _jcatalogI18n = require('jcatalog-i18n');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var formatPatterns = {
  'en': {
    'datePattern': 'MM/dd/yyyy',
    'integerPattern': '#,##0',
    'numberPattern': '#,##0.00',
    'numberDecimalSeparator': '.',
    'numberGroupingSeparator': ',',
    'numberGroupingSeparatorUse': true
  },
  de: {
    datePattern: "dd.MM.yyyy",
    integerPattern: "#,##0",
    numberPattern: "#,##0.00",
    numberDecimalSeparator: ",",
    numberGroupingSeparator: ".",
    numberGroupingSeparatorUse: true
  }
};

var TermReferenceSearchDialogScope = (0, _jcatalogShowroom.showroomScopeDecorator)(_class = function (_Component) {
  _inherits(TermReferenceSearchDialogScope, _Component);

  function TermReferenceSearchDialogScope(props) {
    _classCallCheck(this, TermReferenceSearchDialogScope);

    var _this = _possibleConstructorReturn(this, (TermReferenceSearchDialogScope.__proto__ || Object.getPrototypeOf(TermReferenceSearchDialogScope)).call(this, props));

    _this.state = { showModal: false };
    return _this;
  }

  _createClass(TermReferenceSearchDialogScope, [{
    key: 'getChildContext',
    value: function getChildContext() {
      if (!this.context.i18n) {
        this.context.i18n = new _jcatalogI18n.I18nManager('en', null, formatPatterns);
      }
      return { i18n: this.context.i18n };
    }
  }, {
    key: 'toggleModal',
    value: function toggleModal() {
      this.setState({ showModal: !this.state.showModal });
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          'button',
          { onClick: this.toggleModal.bind(this) },
          'Open modal'
        ),
        this._renderChildren()
      );
    }
  }]);

  return TermReferenceSearchDialogScope;
}(_react.Component)) || _class;

TermReferenceSearchDialogScope.contextTypes = {
  i18n: _react.PropTypes.object
};
TermReferenceSearchDialogScope.childContextTypes = {
  i18n: _react.PropTypes.object
};

exports.default = TermReferenceSearchDialogScope;
