// jcatalog-react-reference-search/lib/components/TermReferenceSearchDialog/utils/termDialogFieldHelpers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getResultFields = exports.getSearchFields = undefined;

var _TranslationsStatus = require('../inputComponents/TranslationsStatus.react');

var _TranslationsStatus2 = _interopRequireDefault(_TranslationsStatus);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getSearchFields(i18n) {
  return [{
    name: 'translations.text',
    label: i18n.getMessage('TermReferenceSearchDialog.translationsText')
  }, {
    name: 'id',
    label: i18n.getMessage('TermReferenceSearchDialog.id')
  }, {
    name: 'comment',
    label: i18n.getMessage('TermReferenceSearchDialog.comment')
  }, {
    name: 'translations.status',
    label: i18n.getMessage('TermReferenceSearchDialog.translationsStatus'),
    inputComponent: _TranslationsStatus2.default
  }];
}

function getResultFields(i18n) {
  return [{
    name: 'translations[0].text',
    label: i18n.getMessage('TermReferenceSearchDialog.translationsText')
  }, {
    name: 'id',
    label: i18n.getMessage('TermReferenceSearchDialog.id'),
    sortable: true,
    rowValueLabel: {
      getValue: function getValue(item) {
        if (item['status'] === 'archived') {
          return item['status'];
        }

        return '';
      },
      className: 'label label-default'
    }

  }, {
    name: 'status',
    label: '',
    sortable: true,
    hidden: true
  }, {
    name: 'comment',
    label: i18n.getMessage('TermReferenceSearchDialog.comment'),
    sortable: true
  }, {
    name: 'translations[0].status',
    label: i18n.getMessage('TermReferenceSearchDialog.translationsStatus')
  }];
}

exports.getSearchFields = getSearchFields;
exports.getResultFields = getResultFields;
