// jcatalog-react-reference-search/lib/components/ClassificationBrowserDialog/ClassificationGroupPreview/ClassificationGroupPreview.react.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _ClassificationGroupPreviewRenderer = require('./partials/ClassificationGroupPreviewRenderer.react');

var _ClassificationGroupPreviewRenderer2 = _interopRequireDefault(_ClassificationGroupPreviewRenderer);

var _i18n = require('./i18n');

var _i18n2 = _interopRequireDefault(_i18n);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ClassificationGroupPreview = (_temp = _class = function (_React$Component) {
  _inherits(ClassificationGroupPreview, _React$Component);

  function ClassificationGroupPreview(props) {
    _classCallCheck(this, ClassificationGroupPreview);

    var _this = _possibleConstructorReturn(this, (ClassificationGroupPreview.__proto__ || Object.getPrototypeOf(ClassificationGroupPreview)).call(this, props));

    _this.state = {
      previewData: null
    };
    return _this;
  }

  _createClass(ClassificationGroupPreview, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.context.i18n.register('ClassificationGroupPreview', _i18n2.default);
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps, nextContext) {
      if (!_lodash2.default.isEqual(nextProps.classificationGroup, this.props.classificationGroup) && !_lodash2.default.isNil(nextProps.classificationGroup)) {
        this._fetchData(nextProps.classificationGroup);
      }
    }
  }, {
    key: '_fetchData',
    value: function _fetchData(classificationGroup) {
      var _this2 = this;

      var _context = this.context,
          classificationService = _context.classificationService,
          attributeService = _context.attributeService,
          domainCoreService = _context.domainCoreService;

      _bluebird2.default.all([domainCoreService.getStatus("CLASSIFICATIONGROUP", classificationGroup.statusId).then(function (response) {
        return { status: response.body };
      }), classificationService.getClassificationGroups(classificationGroup.classification.classificationId, { parentClassificationGroupId: classificationGroup.classificationGroupId }).then(function (response) {
        return { subClassificationGroups: response.body };
      }), classificationService.getClassificationGroupAttributeIds(classificationGroup.classification.classificationId, classificationGroup.classificationGroupId).then(function (response) {
        return attributeService.getAttributes(_lodash2.default.map(response.body, function (attribute) {
          return attribute.attributeId;
        })).then(function (response) {
          return { attributes: response.body };
        });
      })]).then(function (result) {
        var newState = {};
        _lodash2.default.each(result, function (dataObject) {
          _lodash2.default.extend(newState, dataObject);
        });
        _this2.setState(newState);
      }).catch(function (errors) {
        console.error(errors);
      });
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(_ClassificationGroupPreviewRenderer2.default, {
        status: this.state.status,
        attributes: this.state.attributes,
        subClassificationGroups: this.state.subClassificationGroups,
        classificationGroup: this.props.classificationGroup
      });
    }
  }]);

  return ClassificationGroupPreview;
}(_react2.default.Component), _class.propTypes = {
  classificationGroup: _react2.default.PropTypes.object
}, _class.contextTypes = {
  i18n: _react2.default.PropTypes.object.isRequired,
  classificationService: _react2.default.PropTypes.object.isRequired,
  attributeService: _react2.default.PropTypes.object.isRequired,
  domainCoreService: _react2.default.PropTypes.object.isRequired
}, _class.defaultProps = {
  classificationGroup: null
}, _temp);
exports.default = ClassificationGroupPreview;
