// jcatalog-react-reference-search/lib/components/TermInput/TermInput.react.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _superagentBluebirdPromise = require('superagent-bluebird-promise');

var _superagentBluebirdPromise2 = _interopRequireDefault(_superagentBluebirdPromise);

var _querystring = require('querystring');

var _querystring2 = _interopRequireDefault(_querystring);

var _i18n = require('../TermReferenceSearchDialog/i18n');

var _i18n2 = _interopRequireDefault(_i18n);

var _ReactSelectSpecificProps = require('../ReactSelectSpecificProps');

var _ReactSelectSpecificProps2 = _interopRequireDefault(_ReactSelectSpecificProps);

var _ReferenceInputBaseProps = require('../ReferenceInputBaseProps');

var _ReferenceInputBaseProps2 = _interopRequireDefault(_ReferenceInputBaseProps);

var _ReferenceSearchInput = require('../ReferenceSearchInput');

var _ReferenceSearchInput2 = _interopRequireDefault(_ReferenceSearchInput);

var _ReferenceAutocomplete = require('../ReferenceAutocomplete');

var _ReferenceAutocomplete2 = _interopRequireDefault(_ReferenceAutocomplete);

var _termDialogFieldHelpers = require('../TermReferenceSearchDialog/utils/termDialogFieldHelpers');

var _ServiceRegistryValidator = require('../ServiceRegistryValidator');

var _ServiceRegistryValidator2 = _interopRequireDefault(_ServiceRegistryValidator);

var _refreshValueDecorator = require('../../utils/refreshValueDecorator.react');

var _refreshValueDecorator2 = _interopRequireDefault(_refreshValueDecorator);

var _referenceInputUtils = require('../../utils/referenceInputUtils.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SERVICE_NAME = 'term';

var loadTerm = function loadTerm(serviceRegistry) {
  return function (id) {
    return _superagentBluebirdPromise2.default.get(serviceRegistry(SERVICE_NAME).url + '/api/terms/' + id).set('Accept', 'application/json');
  };
};

/**
 * Provides possibility to select term using autocomplete widget or using reference search dialog.
 *
 * @author Eugene Viktarovich
 */
var TermInput = (_dec = (0, _refreshValueDecorator2.default)('id', '_objectLabel', loadTerm), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(TermInput, _Component);

  function TermInput(props, context) {
    _classCallCheck(this, TermInput);

    var _this = _possibleConstructorReturn(this, (TermInput.__proto__ || Object.getPrototypeOf(TermInput)).call(this, props, context));

    _this.context.i18n.register('TermReferenceSearchDialog', _i18n2.default);
    return _this;
  }

  _createClass(TermInput, [{
    key: 'render',
    value: function render() {
      var i18n = this.context.i18n;

      var serviceUrl = this.props.serviceRegistry(SERVICE_NAME).url;

      var referenceSearchProps = _lodash2.default.extend(
      // copy this properties
      _lodash2.default.pick(this.props, ['id', 'name', 'onBlur', 'onFocus', 'onChange', 'multiple', 'disabled', 'readOnly']),
      // add custom properties
      {
        value: this.props.value,
        referenceSearchAction: function referenceSearchAction(searchParams, callback) {
          var queryParams = _querystring2.default.stringify(searchParams);
          return _superagentBluebirdPromise2.default.get(serviceUrl + '/api/terms?' + queryParams).set('Accept', 'application/json').then(function (response) {
            return callback({
              count: (0, _referenceInputUtils.getTotalCount)(response),
              items: response.body
            });
          });
        },
        searchFields: (0, _termDialogFieldHelpers.getSearchFields)(i18n),
        resultFields: (0, _termDialogFieldHelpers.getResultFields)(i18n),

        title: i18n.getMessage('TermReferenceSearchDialog.dialogTitle'),
        labelProperty: '_objectLabel',
        valueProperty: 'id'
      });

      var autocompleteProps = {
        labelProperty: referenceSearchProps.labelProperty,
        valueProperty: referenceSearchProps.valueProperty,
        autocompleteAction: function autocompleteAction(searchTerm /* , callback*/) {
          var queryParams = _querystring2.default.stringify({
            q: searchTerm
          });
          return _superagentBluebirdPromise2.default.get(serviceUrl + '/api/terms?' + queryParams).set('Accept', 'application/json').then(function (response) {
            return {
              options: response.body,
              complete: false
            };
          });
        },
        reactSelectSpecificProps: this.props.reactSelectSpecificProps
      };

      return _react2.default.createElement(
        _ReferenceSearchInput2.default,
        referenceSearchProps,
        _react2.default.createElement(_ReferenceAutocomplete2.default, autocompleteProps)
      );
    }
  }]);

  return TermInput;
}(_react.Component), _class2.propTypes = _extends({}, _ReferenceInputBaseProps2.default, {
  reactSelectSpecificProps: _react2.default.PropTypes.shape(_ReactSelectSpecificProps2.default),
  serviceRegistry: (0, _ServiceRegistryValidator2.default)(SERVICE_NAME)
}), _class2.contextTypes = {
  i18n: _react.PropTypes.object.isRequired
}, _temp)) || _class);
exports.default = TermInput;
